src/sympc/module/nn/functional.py

```diff
@@ -37,3 +49,327 @@ def mse_loss(pred: MPCTensor, target: MPCTensor, reduction: str = "mean") -> MPC
+        kernel_size (Tuple[int, int]): the kernel size
+        stride (Tuple[int, int]): the stride size
+        padding (Tuple[int, int]): the padding size
+        dilation (Tuple[int, int]): the dilation size
+
+    Returns:
+        The reshaped MPCTensor.
+    """
+    session = x.session
+
+    args = [[share, kernel_size, stride, padding, dilation] for share in x.share_ptrs]
+    shares = parallel_execution(helper_max_pool2d_reshape, session.parties)(args)
+
+    res_shape = shares[0].shape.get()
+    res = MPCTensor(shares=shares, session=session, shape=res_shape)
+    return res
+
+
+def helper_max_pool2d_reshape(
+    x: ShareTensor,
+    kernel_size: Tuple[int, int],
+    stride: Tuple[int, int],
+    padding: Tuple[int, int],
+    dilation: Tuple[int, int],
+) -> ShareTensor:
+    """Function that runs at each party for preparing the share.
+
+    Reshape each share tensor to prepare it for calling 'argmax'.
+    The new share would have "each element" as the input on which we
+    will run the max_pool2d kernel.
+
+    Args:
+        x (ShareTensor): the ShareTensor on which to apply the reshaping
+        kernel_size (Tuple[int, int]): the kernel size
+        stride (Tuple[int, int]): the stride size
+        padding (Tuple[int, int]): the padding size
+        dilation (Tuple[int, int]): the dilation size
+
+    Returns:
+        The prepared share tensor (reshaped)
+    """
+    session = get_session(x.session_uuid)
+    tensor = x.tensor.numpy()
+
+    padding = [(0, 0)] * len(tensor.shape[:-2]) + [
+        (padding[0], padding[0]),
+        (padding[1], padding[1]),
+    ]
+    tensor_type = session.tensor_type
+
+    padding_value = 0
+    if session.rank == 0:
+        # ATTENTION: Min value for max_pool2d that works -25
+        padding_value = -25
+
+    padding_value = x.fp_encoder.encode(padding_value)
+    tensor = np.pad(tensor, padding, mode="constant", constant_values=padding_value)
+
+    output_shape = tensor.shape[:-2]
+    output_shape += (
+        (tensor.shape[-2] - kernel_size[0]) // stride[0] + 1,
+        (tensor.shape[-1] - kernel_size[1]) // stride[1] + 1,
+    )
+    output_shape += kernel_size
+
+    output_strides = tensor.strides[:-2]
+    output_strides += (stride[0] * tensor.strides[-2], stride[1] * tensor.strides[-1])
+    output_strides += tensor.strides[-2:]
+
+    window_view_share = torch.tensor(
+        np.lib.stride_tricks.as_strided(
+            tensor, shape=output_shape, strides=output_strides
+        ),
+        dtype=tensor_type,
+    )
+
+    window_view_share = window_view_share.reshape(-1, *kernel_size)
+
+    res_share = ShareTensor(config=x.config)
+    res_share.tensor = window_view_share
+    return res_share
+
+
+def max_pool2d(
+    x: MPCTensor,
+    kernel_size: Union[int, Tuple[int, int]],
+    stride: Optional[Union[int, Tuple[int, int]]] = None,
+    padding: Union[int, Tuple[int, int]] = 0,
+    dilation: Union[int, Tuple[int, int]] = 1,
```

Comment:
Maybe: add argument return indices similar to Pytorch API: https://pytorch.org/docs/stable/generated/torch.nn.MaxPool2d.html

Reply:
I added - only that we can not test it because that parameter should be moved to module (like torch has MaxPool2d). For the moment, we use directly a function -- this one from here (similar to `torch.max_pool2d` - but that one does not have `return_indices` as param).

We also need to implement ```sympc.nn.moduleMaxPool2d``` similar to `torch.nn.module.MaxPool2d` (and this has the parameter `return_indices`)